arch/arm64/src/imx9/imx9_usdhc.c

```diff
@@ -0,0 +1,3364 @@
+    {
+      priv->cintints = USDHC_INT_CINT;
+    }
+  else
+    {
+      priv->cintints = 0;
+    }
+
+#if defined(CONFIG_MMCSD_HAVE_CARDDETECT)
+  if (priv->sw_cd_gpio == 0)
+    {
+      priv->cintints |= USDHC_INT_CINS | USDHC_INT_CRM;
+    }
+#endif
+
+  flags  = enter_critical_section();
+  regval = getreg32(priv->addr + IMX9_USDHC_IRQSIGEN_OFFSET);
+  regval = (regval & ~USDHC_INT_CINT) | priv->cintints;
+  putreg32(regval, priv->addr + IMX9_USDHC_IRQSIGEN_OFFSET);
+  leave_critical_section(flags);
+}
+
+/****************************************************************************
+ * Name: imx9_sdhc_initialize
+ *
+ * Description:
+ *   Initialize SDIO for operation.
+ *
+ * Input Parameters:
+ *   slotno - Slot to be used
+ *
+ * Returned Value:
+ *   A reference to an SDIO interface structure.
+ *   NULL is returned on failures.
+ *
+ ****************************************************************************/
+
+struct sdio_dev_s *imx9_usdhc_initialize(int slotno)
+{
+  DEBUGASSERT(slotno < IMX9_MAX_SDHC_DEV_SLOTS);
+  struct imx9_dev_s *priv = &g_sdhcdev[slotno];
+
+  /* Initialize the USDHC slot structure data structure */
+
+  switch (priv->addr)
+    {
+#if defined(CONFIG_IMX9_USDHC1)
+    case IMX9_USDHC1_BASE:
+
+# if defined(CONFIG_IMX9_USDHC1_WIDTH_D1_D4) || defined(CONFIG_IMX9_USDHC1_WIDTH_D1_D8)
+      imx9_config_gpio(PIN_USDHC1_D1);
+      imx9_iomux_configure(PIN_USDHC1_D1_MUX);
+      imx9_config_gpio(PIN_USDHC1_D2);
+      imx9_iomux_configure(PIN_USDHC1_D2_MUX);
+      imx9_config_gpio(PIN_USDHC1_D3);
+      imx9_iomux_configure(PIN_USDHC1_D3_MUX);
+# endif
+
+# if defined(CONFIG_IMX9_USDHC1_WIDTH_D1_D8)
+      imx9_config_gpio(PIN_USDHC1_D4);
+      imx9_iomux_configure(PIN_USDHC1_D4_MUX);
+      imx9_config_gpio(PIN_USDHC1_D5);
+      imx9_iomux_configure(PIN_USDHC1_D5_MUX);
+      imx9_config_gpio(PIN_USDHC1_D6);
+      imx9_iomux_configure(PIN_USDHC1_D6_MUX);
+      imx9_config_gpio(PIN_USDHC1_D7);
+      imx9_iomux_configure(PIN_USDHC1_D7_MUX);
+# endif
+
+  /* Clocking and CMD pins (all data widths) */
+
+      imx9_config_gpio(PIN_USDHC1_D0);
+      imx9_iomux_configure(PIN_USDHC1_D0_MUX);
+      imx9_config_gpio(PIN_USDHC1_DCLK);
+      imx9_iomux_configure(PIN_USDHC1_DCLK_MUX);
+      imx9_config_gpio(PIN_USDHC1_CMD);
+      imx9_iomux_configure(PIN_USDHC1_CMD_MUX);
+
+# if defined(CONFIG_MMCSD_HAVE_CARDDETECT)
+#   if defined(PIN_USDHC1_CD)
+      imx9_config_gpio(PIN_USDHC1_CD);
+      imx9_iomux_configure(PIN_USDHC1_CD_MUX);
+#   else
+      if (priv->sw_cd_gpio != 0)
+        {
+          imx9_config_gpio(priv->sw_cd_gpio);
+
+        }
+#   endif
+# endif
+
+      /* Enable clocks */
+
+      imx9_ccm_configure_root_clock(CCM_CR_USDHC2, SYS_PLL1PFD1DIV2, 1);
+
+      imx9_ccm_gate_on(CCM_LPCG_USDHC1, true);
+
+      break;
+#endif
+
+#if defined(CONFIG_IMX9_USDHC2)
+    case IMX9_USDHC2_BASE:
+# if defined(CONFIG_IMX9_USDHC2_WIDTH_D1_D4)
+      imx9_config_gpio(PIN_USDHC2_D1);
```

Comment:
Since pins are muxed to SDHC block,there is not point in making GPIO config calls (imx9_config_gpio) ...